### DISCUSSION TITLE
Trying to register additional types in a command interceptor

### DISCUSSION THREAD

**Discussion by 0xced** [spectreconsole/spectre.console#974] [Q&A]
I'm trying to register an `IDbContextFactory<MyContext>` from within a command interceptor to make it available to my commands.

Here's my interceptor:
```c#
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Infrastructure;
using Spectre.Console.Cli;

public class InjectionInterceptor : ICommandInterceptor
{
    private readonly ITypeRegistrarFrontend _registrar;

    public InjectionInterceptor(ITypeRegistrarFrontend registrar)
    {
        _registrar = registrar ?? throw new ArgumentNullException(nameof(registrar));
    }

    public void Intercept(CommandContext context, CommandSettings settings)
    {
        if (settings is DatabaseSettings databaseSettings)
        {
            var dbContextOptions = new DbContextOptionsBuilder<MyContext>().UseSqlServer(databaseSettings.ConnectionString).Options;
            var dbContextFactory = new PooledDbContextFactory<MyContext>(dbContextOptions);
            _registrar.RegisterInstance<IDbContextFactory<MyContext>>(dbContextFactory);
        }
    }
}
```

And here's how I'm using it:
```c#
var app = new CommandApp();
app.Configure(config =>
{
    // ...
    config.SetInterceptor(new InjectionInterceptor(config.Settings.Registrar));
});
return app.Run(args);
```

Unfortunately, this doesn't work because the interceptor is called [after](https://github.com/spectreconsole/spectre.console/blob/088db165edded697f597554ad9d2e71eb1c83acf/src/Spectre.Console.Cli/Internal/CommandExecutor.cs#L98) the [type resolver is built](https://github.com/spectreconsole/spectre.console/blob/088db165edded697f597554ad9d2e71eb1c83acf/src/Spectre.Console.Cli/Internal/CommandExecutor.cs#L73).

To workaround this issue I built an abstract `AsyncCommand<TSettings>` base class with a `protected IDbContextFactory<MyContext> DbContextFactory { get; private set; } = null!;` property initialized at the very beginning of the `ExecuteAsync` method. It's not a bad solution but I feel that injecting the `IDbContextFactory<MyContext>` from the interceptor would have been a cleaner solution.

Is there another way that I could inject my `IDbContextFactory<MyContext>` without a base class? Would it be possible to call the interceptor _before_ the type resolver is built? (It looks like a chicken and egg problem though. 😕)

**Comment by aa-dit-yuh** [comment]
Hi @0xced ,
Any guidance exists today?

**Reply by nils-a** [reply]
No, interceptors are called after the `ITypeResolver` is built.
Unless you have some strange custom implementation of `ITypeRegistrar`/`ITypeResolver`, usage of the registrar is not supported when the resolver has been built.
